transaction.go

```diff
@@ -55,28 +82,149 @@ func (tx *readOnlyTransaction) Rollback() error {
 	return nil
 }
 
-func (tx *readOnlyTransaction) Query(ctx context.Context, stmt spanner.Statement) *spanner.RowIterator {
-	return tx.roTx.Query(ctx, stmt)
+func (tx *readOnlyTransaction) Query(ctx context.Context, stmt spanner.Statement) rowIterator {
+	return &readOnlyRowIterator{tx.roTx.Query(ctx, stmt)}
 }
 
 func (tx *readOnlyTransaction) ExecContext(_ context.Context, stmt spanner.Statement) (int64, error) {
 	return 0, fmt.Errorf("read-only transactions cannot write")
 }
 
+// ErrAbortedDueToConcurrentModification is returned by a read/write transaction
+// that was aborted by Cloud Spanner, and where the internal retry attempt
+// failed because it detected that the results during the retry were different
+// from the initial attempt.
+var ErrAbortedDueToConcurrentModification = status.Error(codes.Aborted, "Transaction was aborted due to a concurrent modification")
+
+// readWriteTransaction is the internal structure for go/sql read/write
+// transactions. These transactions can automatically be retried if the
+// underlying Spanner transaction is aborted. This is done by keeping track
+// of all statements that are executed on the transaction. If the transaction
+// is aborted, the transaction will be replayed using a new read/write
+// transaction on Spanner, and the results of the two will be compared with each
+// other. If they are equal, the underlying Spanner read/write transaction is
+// replaced with the one that was used for the replay, and the user transaction
+// can continue as if nothing happened.
 type readWriteTransaction struct {
-	ctx   context.Context
+	ctx    context.Context
+	client *spanner.Client
+	// rwTx is the underlying Spanner read/write transaction. This transaction
+	// will be replaced with a new one if the initial transaction is aborted.
 	rwTx  *spanner.ReadWriteStmtBasedTransaction
 	close func()
+	// retryAborts indicates whether this transaction will automatically retry
+	// the transaction if it is aborted by Spanner. The default is true.
+	retryAborts bool
+
+	// statements contains the list of statements that has been executed on this
+	// transaction so far. These statements will be replayed on a new r/w tx if
```

Comment:
To be consistent. `r/w tx` -> `read/write transaction`

Reply:
done